src/js/components/Button/Button.js

```diff
@@ -258,6 +331,70 @@ const Button = forwardRef(
       contents = first || second || children;
     }
 
+    const defaultBadgeDimension =
+      typeof badgeProp === 'boolean' ||
+      (badgeProp && badgeProp.value && typeof badgeProp.value === 'boolean')
+        ? // empty badge should be smaller. this value was chosen as a default
+          // after experimenting with various values
+          `${parseInt(theme.button.badge.size.medium.replace('px', ''), 10) /
+            2}px`
+        : theme.button.badge.size.medium;
+
+    const [badgeWidth, setBadgeWidth] = useState(defaultBadgeDimension);
+    // scale badge to fit its contents
+    // width informs how far to horizontally offset the badge
+    useEffect(() => {
+      if (badgeRef && badgeRef.current && typeof badgeProp !== 'boolean') {
+        setBadgeWidth(getBadgeDimension('width', badgeProp, badgeRef, theme));
+      }
+    }, [badgeProp, theme]);
+
+    const [badgeHeight, setBadgeHeight] = useState(defaultBadgeDimension);
+    // height informs how far to vertically offset the badge
+    useEffect(() => {
+      if (badgeRef && badgeRef.current && typeof badgeProp !== 'boolean')
+        setBadgeHeight(getBadgeDimension('height', badgeProp, badgeRef, theme));
+    }, [badgeProp, theme]);
+
+    // offset the badge so it overlaps content. when badge has content,
+    // offset should be 50%. when badge is empty, offset by a smaller amount to
+    // keep the badge closer to the content. this value was chosen as a
+    // reasonable default after testing with various grommet icons.
+    const divisor =
+      typeof badgeProp === 'boolean' || (badgeProp && badgeProp.value === true)
+        ? 3.5
+        : 2;
+    const verticalOffset = `-${parseInt(badgeHeight.replace('px', ''), 10) /
+      divisor}px`;
+    const horizontalOffset = `-${parseInt(badgeWidth.replace('px', ''), 10) /
+      divisor}px`;
+
+    // set the badge relative to the button content
+    // as opposed to outer edge of button
+    if (badgeProp && badgeProp.target === 'contents') {
```

Comment:
Would it be possible to do this more automatically by checking if the Button has a border or background to decide where to align to?

Reply:
Applied this change in recent commit.